### ISSUE TITLE
threads::shared: cannot assign anonymous data refs to shared scalar

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5602]
Migrated from [rt.perl.org#9748](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=9748) (status was 'resolved')

Searchable as RT9748$

**Comment by p5pRT** [comment]
### From perl58@dev1.ridas.co.uk

#### Created by scriptyrich@yahoo.co.uk



&nbsp;  use threads;
&nbsp;  use threads&#8203;::shared;

&nbsp;  my $array\_ref : shared = \[1\, 2\, 3\, 4\];

Gives&#8203;:

&nbsp;  Invalid value for shared scalar at e\.pl line 4\.

Also true for \{\}\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.0:

Configured by root at Sun Jun 16 00:35:43 BST 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.18-6mdk, archname=i686-linux-thread-multi
    uname='linux dev1.ridas.co.uk 2.4.18-6mdk #1 fri mar 15 02:59:08 cet 2002 i686 unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='2.96 20000731 (Mandrake Linux 8.2 2.96-0.76mdk)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=/lib/libc-2.2.4.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/local/perl/devel/5.8/lib/5.8.0/i686-linux-thread-multi
    /usr/local/perl/devel/5.8/lib/5.8.0
    /usr/local/perl/devel/5.8/lib/site_perl/5.8.0/i686-linux-thread-multi
    /usr/local/perl/devel/5.8/lib/site_perl/5.8.0
    /usr/local/perl/devel/5.8/lib/site_perl
    .


Environment for perl v5.8.0:
    HOME=/home/perl58
    LANG=en_GB
    LANGUAGE=en_GB:en
    LC_COLLATE=en_GB
    LC_CTYPE=en_GB
    LC_MESSAGES=en_GB
    LC_MONETARY=en_GB
    LC_NUMERIC=en_GB
    LC_TIME=en_GB
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/perl/devel/5.8/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/local/bin:/usr/games:/usr/local/java/blackdown/j2sdk1.3.1/bin:/home/rich/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Jun 18\, 2002 at 07&#8203;:24&#8203;:51PM \+0100\, Used for testing Perl 5\.8 wrote&#8203;:
>   use threads;
>   use threads&#8203;::shared;
> 
>   my $array\_ref : shared = \[1\, 2\, 3\, 4\];
> 
> Gives&#8203;:
> 
>   Invalid value for shared scalar at e\.pl line 4\.
> 
> Also true for \{\}\.

This is the correct behaviour\. A shared scalar cannot contain a reference to
private data\. You probably want to do something like the following instead&#8203;:

&nbsp;    my @&#8203;array : shared = \(1\,2\,3\,4\);
&nbsp;    my $array\_ref : shared = \\@&#8203;array;

Dave\.

\-\- 
"Strange women lying in ponds distributing swords is no basis for a system
of government\. Supreme executive power derives from a mandate from the
masses\, not from some farcical aquatic ceremony\."
Dennis \- Monty Python and the Holy Grail\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]


On måndag\, juni 24\, 2002\, at 12&#8203;:54 \, Dave Mitchell wrote&#8203;:

> On Tue\, Jun 18\, 2002 at 07&#8203;:24&#8203;:51PM \+0100\, Used for testing Perl 5\.8 
> wrote&#8203;:
>>   use threads;
>>   use threads&#8203;::shared;
>>
>>   my $array\_ref : shared = \[1\, 2\, 3\, 4\];
>>
>> Gives&#8203;:
>>
>>   Invalid value for shared scalar at e\.pl line 4\.
>>
>> Also true for \{\}\.
>
> This is the correct behaviour\. A shared scalar cannot contain a 
> reference to
> private data\. You probably want to do something like the following 
> instead&#8203;:
>
>     my @&#8203;array : shared = \(1\,2\,3\,4\);
>     my $array\_ref : shared = \\@&#8203;array;
>
> Dave\.
>
> \-\-
> "Strange women lying in ponds distributing swords is no basis for a 
> system
> of government\. Supreme executive power derives from a mandate from the
> masses\, not from some farcical aquatic ceremony\."
> Dennis \- Monty Python and the Holy Grail\.
>


Bleh\, seems like prototypes really suck\,

\\\[$@&#8203;%\] will not accept  share\(\[\]\); for some strange stupid reason\, seems 
like the synopsis needs to be changed :\-\(

Arthur

**Comment by p5pRT** [comment]
### From arthur@contiller.se

Hello\,

This was fixed with change \#17453 so you can now write

my $foo = &shared\(\[1\,2\,3\]\);

Arthur

**Comment by p5pRT** [comment]
arthur@contiller.se - Status changed from 'open' to 'resolved'
